finality-aleph/src/sync/forest/mod.rs

```diff
@@ -0,0 +1,721 @@
+use std::collections::{
+    hash_map::{Entry, OccupiedEntry, VacantEntry},
+    HashMap, HashSet,
+};
+
+use crate::sync::{BlockIdentifier, Header, Justification, PeerId};
+
+mod vertex;
+
+use vertex::{JustificationAddResult, Vertex};
+
+type BlockIdFor<J> = <<J as Justification>::Header as Header>::Identifier;
+
+pub struct JustificationWithParent<J: Justification> {
+    pub justification: J,
+    pub parent: BlockIdFor<J>,
+}
+
+enum VertexHandle<'a, I: PeerId, J: Justification> {
+    HopelessFork,
+    BelowMinimal,
+    HighestFinalized,
+    Unknown(VacantEntry<'a, BlockIdFor<J>, VertexWithChildren<I, J>>),
+    Candidate(OccupiedEntry<'a, BlockIdFor<J>, VertexWithChildren<I, J>>),
+}
+
+/// Our interest in a block referred to by a vertex, including the information about whom we expect to have the block.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Interest<I: PeerId> {
+    /// We are not interested in this block.
+    Uninterested,
+    /// We would like to have this block.
+    Required(HashSet<I>),
+    /// We would like to have this block and its the highest on its branch.
+    TopRequired(HashSet<I>),
+}
+
+/// What can go wrong when inserting data into the forest.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    HeaderMissingParentId,
+    IncorrectParentState,
+    IncorrectVertexState,
+    ParentNotImported,
+    TooNew,
+}
+
+pub struct VertexWithChildren<I: PeerId, J: Justification> {
+    vertex: Vertex<I, J>,
+    children: HashSet<BlockIdFor<J>>,
+}
+
+impl<I: PeerId, J: Justification> VertexWithChildren<I, J> {
+    fn new() -> Self {
+        Self {
+            vertex: Vertex::new(),
+            children: HashSet::new(),
+        }
+    }
+
+    fn add_child(&mut self, child: BlockIdFor<J>) {
+        self.children.insert(child);
+    }
+}
+
+// How deep can the forest be, vaguely based on two sessions ahead, which is the most we expect to
+// ever need worst case scenario.
+const MAX_DEPTH: u32 = 1800;
```

Comment:
nit: this could be somehow dependent on `DEFAULT_SESSION_PERIOD` from `aleph_primitives.rs` in final implementation of sync. So this could be an argument for forest struct.

Reply:
Hmm, technically the reason for this constant and the reasoning for it's value are separate – I think it's fine that we have it be 1800 to make sure the recursion is not too cumbersome whatever the session period is?